File: bindings/ruby/test/interaction_test.rb
# frozen_string_literal: true

# Interaction Tests - Parser → Compiler → Emitter handoffs
#
# This test suite validates the handoff between pipeline stages:
# - Parser → Compiler: Ensures AST is correctly consumed
# - Compiler → Emitter: Ensures IR is correctly transformed to regex

require 'minitest/autorun'
require_relative '../lib/strling/core/parser'
require_relative '../lib/strling/core/compiler'
require_relative '../lib/strling/emitters/pcre2'

class InteractionTest < Minitest::Test

  # ============================================================================
  # Parser → Compiler Handoff Tests
  # ============================================================================

  def test_parser_compiler_simple_literal
    parser = Strling::Core::Parser.new("hello")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)

    assert_not_nil ir
    serialized = serialize(ir)
    assert_equal 'Lit', serialized['ir']
  end

  def test_parser_compiler_quantifier
    parser = Strling::Core::Parser.new("a+")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Quant', serialized['ir']
  end

  def test_parser_compiler_character_class
    parser = Strling::Core::Parser.new("[abc]")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'CharClass', serialized['ir']
  end

  def test_parser_compiler_capturing_group
    parser = Strling::Core::Parser.new("(abc)")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Group', serialized['ir']
  end

  def test_parser_compiler_alternation
    parser = Strling::Core::Parser.new("a|b")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Alt', serialized['ir']
  end

  def test_parser_compiler_named_group
    parser = Strling::Core::Parser.new("(?<name>abc)")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Group', serialized['ir']
  end

  def test_parser_compiler_lookahead
    parser = Strling::Core::Parser.new("(?=abc)")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Look', serialized['ir']
  end

  def test_parser_compiler_lookbehind
    parser = Strling::Core::Parser.new("(?<=abc)")
    flags, ast = parser.parse

    ir = Strling::Core::Compiler.compile(ast)
    serialized = serialize(ir)

    assert_equal 'Look', serialized['ir']
  end

  # ============================================================================
  # Compiler → Emitter Handoff Tests
  # ============================================================================

  def test_compiler_emitter_simple_literal
    regex = compile_to_regex("hello")
    assert_equal "hello", regex
  end

  def test_compiler_emitter_digit_shorthand
    regex = compile_to_regex("\\d+")
    assert_equal "\\d+", regex
  end

  def test_compiler_emitter_character_class
    regex = compile_to_regex("[abc]")
    assert_equal "[abc]", regex
  end

  def test_compiler_emitter_character_class_range
    regex = compile_to_regex("[a-z]")
    assert_equal "[a-z]", regex
  end

  def test_compiler_emitter_negated_class
    regex = compile_to_regex("[^abc]")
    assert_equal "[^abc]", regex
  end

  def test_compiler_emitter_quantifier_plus
    regex = compile_to_regex("a+")
    assert_equal "a+", regex
  end

  def test_compiler_emitter_quantifier_star
    regex = compile_to_regex("a*")
    assert_equal "a*", regex
  end

  def test_compiler_emitter_quantifier_optional
    regex = compile_to_regex("a?")
    assert_equal "a?", regex
  end

  def test_compiler_emitter_quantifier_exact
    regex = compile_to_regex("a{3}")
    assert_equal "a{3}", regex
  end

  def test_compiler_emitter_quantifier_range
    regex = compile_to_regex("a{2,5}")
    assert_equal "a{2,5}", regex
  end

  def test_compiler_emitter_quantifier_lazy
    regex = compile_to_regex("a+?")
    assert_equal "a+?", regex
  end

  def test_compiler_emitter_capturing_group
    regex = compile_to_regex("(abc)")
    assert_equal "(abc)", regex
  end

  def test_compiler_emitter_non_capturing_group
    regex = compile_to_regex("(?:abc)")
    assert_equal "(?:abc)", regex
  end

  def test_compiler_emitter_named_group
    regex = compile_to_regex("(?<name>abc)")
    assert_equal "(?<name>abc)", regex
  end

  def test_compiler_emitter_alternation
    regex = compile_to_regex("cat|dog")
    assert_equal "cat|dog", regex
  end

  def test_compiler_emitter_anchors
    regex = compile_to_regex("^abc$")
    assert_equal "^abc$", regex
  end

  def test_compiler_emitter_positive_lookahead
    regex = compile_to_regex("foo(?=bar)")
    assert_equal "foo(?=bar)", regex
  end

  def test_compiler_emitter_negative_lookahead
    regex = compile_to_regex("foo(?!bar)")
    assert_equal "foo(?!bar)", regex
  end

  def test_compiler_emitter_positive_lookbehind
    regex = compile_to_regex("(?<=foo)bar")
    assert_equal "(?<=foo)bar", regex
  end

  def test_compiler_emitter_negative_lookbehind
    regex = compile_to_regex("(?<!foo)bar")
    assert_equal "(?<!foo)bar", regex
  end

  # ============================================================================
  # Semantic Edge Case Tests
  # ============================================================================

  def test_semantic_duplicate_capture_group
    assert_raises(Strling::Core::STRlingParseError) do
      parser = Strling::Core::Parser.new("(?<name>a)(?<name>b)")
      parser.parse
    end
  end

  def test_semantic_ranges
    # Invalid range [z-a] should produce an error
    assert_raises(Strling::Core::STRlingParseError) do
      parser = Strling::Core::Parser.new("[z-a]")
      parser.parse
    end
  end

  # ============================================================================
  # Full Pipeline Tests
  # ============================================================================

  def test_full_pipeline_phone_number
    regex = compile_to_regex("(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})")
    assert_equal "(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})", regex
  end

  def test_full_pipeline_ipv4
    regex = compile_to_regex("(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})")
    assert_equal "(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})", regex
  end

  private

  def compile_to_regex(dsl)
    parser = Strling::Core::Parser.new(dsl)
    flags, ast = parser.parse
    ir = Strling::Core::Compiler.compile(ast)
    Strling::Emitters::Pcre2.emit(ir, flags)
  end

  def serialize(obj)
    case obj
    when Data
      obj.to_h.transform_keys(&:to_s).transform_values { |v| serialize(v) }.compact
    when Array
      obj.map { |v| serialize(v) }
    when Hash
      obj.transform_keys(&:to_s).transform_values { |v| serialize(v) }.compact
    else
      obj
    end
  end
end

File: bindings/php/tests/InteractionTest.php
<?php

namespace STRling\Tests;

use PHPUnit\Framework\TestCase;
use STRling\Core\Parser;
use STRling\Compiler;
use STRling\Emitters\Pcre2Emitter;
use STRling\Core\Nodes\Flags;

/**
 * Interaction Tests - Parser → Compiler → Emitter handoffs
 *
 * This test suite validates the handoff between pipeline stages:
 * - Parser → Compiler: Ensures AST is correctly consumed
 * - Compiler → Emitter: Ensures IR is correctly transformed to regex
 */
class InteractionTest extends TestCase
{
    private Compiler $compiler;
    private Pcre2Emitter $emitter;

    protected function setUp(): void
    {
        $this->compiler = new Compiler();
        $this->emitter = new Pcre2Emitter();
    }

    // ========================================================================
    // Parser → Compiler Handoff Tests
    // ========================================================================

    public function testParserCompiler_SimpleLiteral(): void
    {
        $parser = new Parser("hello");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $this->assertNotNull($ir);
        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Lit', $serialized['ir']);
    }

    public function testParserCompiler_Quantifier(): void
    {
        $parser = new Parser("a+");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Quant', $serialized['ir']);
    }

    public function testParserCompiler_CharacterClass(): void
    {
        $parser = new Parser("[abc]");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('CharClass', $serialized['ir']);
    }

    public function testParserCompiler_CapturingGroup(): void
    {
        $parser = new Parser("(abc)");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Group', $serialized['ir']);
    }

    public function testParserCompiler_Alternation(): void
    {
        $parser = new Parser("a|b");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Alt', $serialized['ir']);
    }

    public function testParserCompiler_NamedGroup(): void
    {
        $parser = new Parser("(?<name>abc)");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Group', $serialized['ir']);
    }

    public function testParserCompiler_Lookahead(): void
    {
        $parser = new Parser("(?=abc)");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Look', $serialized['ir']);
    }

    public function testParserCompiler_Lookbehind(): void
    {
        $parser = new Parser("(?<=abc)");
        [$flags, $ast] = $parser->parse();

        $ir = $this->compiler->compile($ast);

        $serialized = json_decode(json_encode($ir), true);
        $this->assertEquals('Look', $serialized['ir']);
    }

    // ========================================================================
    // Compiler → Emitter Handoff Tests
    // ========================================================================

    public function testCompilerEmitter_SimpleLiteral(): void
    {
        $regex = $this->compileToRegex("hello");
        $this->assertEquals("hello", $regex);
    }

    public function testCompilerEmitter_DigitShorthand(): void
    {
        $regex = $this->compileToRegex("\\d+");
        $this->assertEquals("\\d+", $regex);
    }

    public function testCompilerEmitter_CharacterClass(): void
    {
        $regex = $this->compileToRegex("[abc]");
        $this->assertEquals("[abc]", $regex);
    }

    public function testCompilerEmitter_CharacterClassRange(): void
    {
        $regex = $this->compileToRegex("[a-z]");
        $this->assertEquals("[a-z]", $regex);
    }

    public function testCompilerEmitter_NegatedClass(): void
    {
        $regex = $this->compileToRegex("[^abc]");
        $this->assertEquals("[^abc]", $regex);
    }

    public function testCompilerEmitter_QuantifierPlus(): void
    {
        $regex = $this->compileToRegex("a+");
        $this->assertEquals("a+", $regex);
    }

    public function testCompilerEmitter_QuantifierStar(): void
    {
        $regex = $this->compileToRegex("a*");
        $this->assertEquals("a*", $regex);
    }

    public function testCompilerEmitter_QuantifierOptional(): void
    {
        $regex = $this->compileToRegex("a?");
        $this->assertEquals("a?", $regex);
    }

    public function testCompilerEmitter_QuantifierExact(): void
    {
        $regex = $this->compileToRegex("a{3}");
        $this->assertEquals("a{3}", $regex);
    }

    public function testCompilerEmitter_QuantifierRange(): void
    {
        $regex = $this->compileToRegex("a{2,5}");
        $this->assertEquals("a{2,5}", $regex);
    }

    public function testCompilerEmitter_QuantifierLazy(): void
    {
        $regex = $this->compileToRegex("a+?");
        $this->assertEquals("a+?", $regex);
    }

    public function testCompilerEmitter_CapturingGroup(): void
    {
        $regex = $this->compileToRegex("(abc)");
        $this->assertEquals("(abc)", $regex);
    }

    public function testCompilerEmitter_NonCapturingGroup(): void
    {
        $regex = $this->compileToRegex("(?:abc)");
        $this->assertEquals("(?:abc)", $regex);
    }

    public function testCompilerEmitter_NamedGroup(): void
    {
        $regex = $this->compileToRegex("(?<name>abc)");
        $this->assertEquals("(?<name>abc)", $regex);
    }

    public function testCompilerEmitter_Alternation(): void
    {
        $regex = $this->compileToRegex("cat|dog");
        $this->assertEquals("cat|dog", $regex);
    }

    public function testCompilerEmitter_Anchors(): void
    {
        $regex = $this->compileToRegex("^abc\$");
        $this->assertEquals("^abc\$", $regex);
    }

    public function testCompilerEmitter_PositiveLookahead(): void
    {
        $regex = $this->compileToRegex("foo(?=bar)");
        $this->assertEquals("foo(?=bar)", $regex);
    }

    public function testCompilerEmitter_NegativeLookahead(): void
    {
        $regex = $this->compileToRegex("foo(?!bar)");
        $this->assertEquals("foo(?!bar)", $regex);
    }

    public function testCompilerEmitter_PositiveLookbehind(): void
    {
        $regex = $this->compileToRegex("(?<=foo)bar");
        $this->assertEquals("(?<=foo)bar", $regex);
    }

    public function testCompilerEmitter_NegativeLookbehind(): void
    {
        $regex = $this->compileToRegex("(?<!foo)bar");
        $this->assertEquals("(?<!foo)bar", $regex);
    }

    // ========================================================================
    // Semantic Edge Case Tests
    // ========================================================================

    public function test_semantic_duplicate_capture_group(): void
    {
        $this->expectException(\STRling\Core\STRlingParseError::class);
        $parser = new Parser("(?<name>a)(?<name>b)");
        $parser->parse();
    }

    public function test_semantic_ranges(): void
    {
        // Invalid range [z-a] should produce an error
        $this->expectException(\STRling\Core\STRlingParseError::class);
        $parser = new Parser("[z-a]");
        $parser->parse();
    }

    // ========================================================================
    // Full Pipeline Tests
    // ========================================================================

    public function testFullPipeline_PhoneNumber(): void
    {
        $regex = $this->compileToRegex("(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})");
        $this->assertEquals("(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})", $regex);
    }

    public function testFullPipeline_IPv4(): void
    {
        $regex = $this->compileToRegex("(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})");
        $this->assertEquals("(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})", $regex);
    }

    // ========================================================================
    // Helper Methods
    // ========================================================================

    private function compileToRegex(string $dsl): string
    {
        $parser = new Parser($dsl);
        [$flags, $ast] = $parser->parse();
        $ir = $this->compiler->compile($ast);
        return $this->emitter->emit($ir, $flags);
    }
}
